### PR TITLE
Public ip prefix

### LINKED ISSUE #2239: Support Public IP Prefixes
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The Azure terraform provider does not support Azure Public IP Prefixes. Please support them. This would mean:

- I can create azure public ip prefixes
- I can create azurerm public ip's (standard) referencing the azure public ip prefix created before

This make on premise firewalling for larger clusters with a lot of NAT (e.g. Kubernetes/Openshift) easier.

### New or Affected Resource(s)

* azurerm_public_ip_prefix
* azurerm_public_ip (modified, add support to create from azurerm_public_ip_prefix)

### References

* See https://docs.microsoft.com/en-us/azure/virtual-network/manage-public-ip-address-prefix

### PR DESCRIPTION
**Pull request by steve-hawkins**
This should fulfil Issue #2239 

(fixes #2239)

### CODE REVIEW

File: azurerm/resource_arm_public_ip_prefix.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+func resourceArmPublicIpPrefix() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmPublicIpPrefixCreateUpdate,
+		Read:   resourceArmPublicIpPrefixRead,
+		Update: resourceArmPublicIpPrefixCreateUpdate,
+		Delete: resourceArmPublicIpPrefixDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```
Comment (katbyte) [inline azurerm/resource_arm_public_ip_prefix.go]: I'm not sure if this is required? The read function should be doing the same check on the `id` passed into it?

File: azurerm/resource_arm_public_ip_prefix.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+			"location": locationSchema(),
+
+			"resource_group_name": resourceGroupNameSchema(),
+
+			"sku": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				ForceNew:         true,
+				Default:          string(network.Standard),
+				DiffSuppressFunc: suppress.CaseDifference,
```
Comment (katbyte) [inline azurerm/resource_arm_public_ip_prefix.go]: We are moving to case sensitivity for properties like this, as such could we more this line?

File: azurerm/resource_arm_public_ip_prefix.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+				ValidateFunc: validation.StringInSlice([]string{
+					string(network.Standard),
+				}, true),
```
Comment (katbyte) [inline azurerm/resource_arm_public_ip_prefix.go]: and set this to false

File: website/docs/r/public_ip_prefix.html.markdown [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name of the Public IP resource . Changing this forces a
+    new resource to be created.
```
Comment (katbyte) [inline website/docs/r/public_ip_prefix.html.markdown]: COuld we put this all on one line?

File: website/docs/r/public_ip_prefix.html.markdown [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+* `resource_group_name` - (Required) The name of the resource group in which to
+    create the public IP.
```
Comment (katbyte) [inline website/docs/r/public_ip_prefix.html.markdown]: COuld we put this all on one line?

File: website/docs/r/public_ip_prefix.html.markdown [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+* `zones` - (Optional) A collection containing the availability zone to allocate the Public IP in.
+
+-> **Please Note**: Availability Zones are [only supported in several regions at this time](https://docs.microsoft.com/en-us/azure/availability-zones/az-overview).
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The Public IP Prefix ID.
+* `ip_prefix` - The IP address prefix value that was allocated.
```
Comment (katbyte) [inline website/docs/r/public_ip_prefix.html.markdown]: COuld we add the import section detailing the import command here?

File: azurerm/resource_arm_public_ip.go [hunk 7]
```diff
@@ -126,6 +126,13 @@ func resourceArmPublicIp() *schema.Resource {
 				Computed: true,
 			},
 
+			"public_ip_prefix_id": {
+				Type:         schema.TypeString,
+				Required:     true,
```
[the remark this answers is not in the document]
Comment (steve-hawkins) [reply azurerm/resource_arm_public_ip.go]: good spot

File: azurerm/resource_arm_public_ip_prefix.go (hunk 1 above)
Comment (steve-hawkins) [reply azurerm/resource_arm_public_ip_prefix.go]: I have to admit this was a case of copy and paste from the azurerm_public_ip resource and I wasn't entirely sure why it was there.

Removed now

**Comment by steve-hawkins** [comment]
@katbyte thanks for your review and comments

hopefully covered them all off now, let me know if all looks good

**Comment by steve-hawkins** [comment]
@katbyte let me know if there is anything else needed to get this PR merged in for 1.24

**Comment by steve-hawkins** [comment]
Hi @katbyte apologies for the delay

Those tests should pass now

**Pull request merged**